primitives/teerex/src/lib.rs

```diff
@@ -45,10 +56,139 @@ impl<PubKey, Url> Enclave<PubKey, Url> {
 	}
 }
 
+/// The list of valid TCBs for an enclave.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, sp_core::RuntimeDebug, TypeInfo)]
+pub struct QeTcb {
+	pub isvsvn: u16,
+}
+
+impl QeTcb {
+	pub fn new(isvsvn: u16) -> Self {
+		Self { isvsvn }
+	}
+}
+
+/// This represents all the collateral data that we need to store on chain in order to verify
+/// the quoting enclave validity of another enclave that wants to register itself on chain
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, sp_core::RuntimeDebug, TypeInfo)]
+pub struct QuotingEnclave {
+	// Todo: make timestamp: Moment
+	pub issue_date: u64, // unix epoch in milliseconds
+	// Todo: make timestamp: Moment
+	pub next_update: u64, // unix epoch in milliseconds
+	pub miscselect: [u8; 4],
+	pub miscselect_mask: [u8; 4],
+	pub attributes: [u8; 16],
+	pub attributes_mask: [u8; 16],
+	pub mrsigner: [u8; 32],
+	pub isvprodid: u16,
+	/// Contains only the TCB versions that are considered UpToDate
+	pub tcb: Vec<QeTcb>,
+}
+
+impl QuotingEnclave {
+	#[allow(clippy::too_many_arguments)]
+	pub fn new(
+		issue_date: u64,
+		next_update: u64,
+		miscselect: [u8; 4],
+		miscselect_mask: [u8; 4],
+		attributes: [u8; 16],
+		attributes_mask: [u8; 16],
+		mrsigner: [u8; 32],
+		isvprodid: u16,
+		tcb: Vec<QeTcb>,
+	) -> Self {
+		Self {
+			issue_date,
+			next_update,
+			miscselect,
+			miscselect_mask,
+			attributes,
+			attributes_mask,
+			mrsigner,
+			isvprodid,
+			tcb,
+		}
+	}
+}
+
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, sp_core::RuntimeDebug, TypeInfo)]
+pub struct TcbVersionStatus {
+	pub cpusvn: Cpusvn,
+	pub pcesvn: Pcesvn,
+}
+
+impl TcbVersionStatus {
+	pub fn new(cpusvn: Cpusvn, pcesvn: Pcesvn) -> Self {
+		Self { cpusvn, pcesvn }
+	}
+
+	pub fn is_valid(&self, examinee: &TcbVersionStatus) -> bool {
+		for (v, r) in self.cpusvn.iter().zip(examinee.cpusvn.iter()) {
+			if *v > *r {
+				return false
+			}
+		}
+		self.pcesvn <= examinee.pcesvn
+	}
+}
+
+/// This represents all the collateral data that we need to store on chain in order to verify
+/// the quoting enclave validity of another enclave that wants to register itself on chain
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, sp_core::RuntimeDebug, TypeInfo)]
+pub struct TcbInfoOnChain {
+	// Todo: make timestamp: Moment
+	pub issue_date: u64, // unix epoch in milliseconds
+	// Todo: make timestamp: Moment
+	pub next_update: u64, // unix epoch in milliseconds
+	tcb_levels: Vec<TcbVersionStatus>,
+}
+
+impl TcbInfoOnChain {
+	pub fn new(issue_date: u64, next_update: u64, tcb_levels: Vec<TcbVersionStatus>) -> Self {
+		Self { issue_date, next_update, tcb_levels }
+	}
+
+	pub fn is_valid(&self, examinee: &TcbVersionStatus) -> Option<TcbVersionStatus> {
```

Comment:
Above you have defined the functino slightly different: `fn is_valid() -> bool` and here it returns an option. Is there any reason for that?

Reply:
I think it might be that the onchain _database_ may not contain the `examinee`, is that so @Niederb?

Reply:
Actually the `Option` was really not needed as it was only checked that it is `Some`. I changed the function to return `bool`. I also renamed both functions to `verify_examinee(..)` to emphasize the fact that it is the parameter that is being verified (and not `Self`). I hope this makes it a bit clearer.

Reply:
Perfect!